controllers/ingress/group_controller.go

```
@@ -116,40 +126,57 @@ type groupReconciler struct {
+	r.reconcileCounters.IncrementIngress(req.NamespacedName)
 	return runtime.HandleReconcileError(r.reconcile(ctx, req), r.logger)
 }
 
 func (r *groupReconciler) reconcile(ctx context.Context, req reconcile.Request) error {
 	ingGroupID := ingress.DecodeGroupIDFromReconcileRequest(req)
-	ingGroup, err := r.groupLoader.Load(ctx, ingGroupID)
+	var err error
+	var ingGroup ingress.Group
+	loadIngressFn := func() {
+		ingGroup, err = r.groupLoader.Load(ctx, ingGroupID)
+	}
+	r.metricsCollector.ObserveControllerReconcileLatency("ingress", "fetch_ingress", loadIngressFn)
 	if err != nil {
-		return err
+		return errmetrics.NewErrorWithMetrics("ingress", "fetch_ingress_error", err, r.metricsCollector)
 	}
 
-	if err := r.groupFinalizerManager.AddGroupFinalizer(ctx, ingGroupID, ingGroup.Members); err != nil {
+	addFinalizerFn := func() {
+		err = r.groupFinalizerManager.AddGroupFinalizer(ctx, ingGroupID, ingGroup.Members)
+	}
+	r.metricsCollector.ObserveControllerReconcileLatency("ingress", "add_group_finalizer", addFinalizerFn)
+	if err != nil {
 		r.recordIngressGroupEvent(ctx, ingGroup, corev1.EventTypeWarning, k8s.IngressEventReasonFailedAddFinalizer, fmt.Sprintf("Failed add finalizer due to %v", err))
-		return err
+		return errmetrics.NewErrorWithMetrics("ingress", "add_group_finalizer_error", err, r.metricsCollector)
 	}
+
 	_, lb, err := r.buildAndDeployModel(ctx, ingGroup)
 	if err != nil {
 		return err
 	}
 
 	if len(ingGroup.Members) > 0 && lb != nil {
-		lbDNS, err := lb.DNSName().Resolve(ctx)
-		if err != nil {
-			return err
-		}
-		if err := r.updateIngressGroupStatus(ctx, ingGroup, lbDNS); err != nil {
-			r.recordIngressGroupEvent(ctx, ingGroup, corev1.EventTypeWarning, k8s.IngressEventReasonFailedUpdateStatus, fmt.Sprintf("Failed update status due to %v", err))
-			return err
+		dnsResolveAndUpdateStatus := func() {
+			lbDNS, err := lb.DNSName().Resolve(ctx)
+			if err != nil {
+				return
```

I believe you still need to propagate this error to the parent function. Right now, it will silently fail.